text/3519-arbitrary-self-types-v2.md

```diff
@@ -19,7 +19,57 @@ This RFC proposes some changes to the existing nightly feature based on the expe
 
 ## Motivation for the arbitrary self types feature overall
 
-One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference.
+One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference. With "arbitrary self types", smart pointer types can be created which obey foreign-language semantics and can be used in safe Rust code:
+
+```rust
+#[repr(transparent)]
+#[derive(Clone)]
+/// A C++ reference. Obeys C++ reference semantics, not Rust reference semantics.
+/// There is no exclusivity; the underlying data may mutate, etc.
```

Comment:
I assume this *does* carry an invariant that the pointer points to dereferenceable memory though and holds a valid instance of the corresponding C++ type? Would be good to document explicitly what is guaranteed about the fields of this type.

Reply:
Maybe the semantics are just as you describe, but the point of this RFC is that _Rust doesn't care_. People should be free to implement whatever semantics they want for foreign language smart pointers.

Rust only starts to care if we dereference the data in Rust. Otherwise, the underlying pointer is just an opaque token which ephemerally passes through the Rust language, and the other language can use it however it wants to.

Reply:
This is an example though. Everyone writing such a type should document their particular invariant, so if the example doesn't do that then it does not follow best practices.

Reply:
I've added a sentence explaining that (for the purposes of this RFC) the invariants are not fully documented.

Actually documenting them would distract from this RFC, because pointer and reference semantics in C++ are essentially unpredictable in a sufficiently complex codebase. There's a reasonable chance that this pointer could point to the null page, or to a deallocated object, or just to something entirely wrong. That's the inevitability of C++, and in a sufficiently complex interop scenario, there's every chance that those problems end up applying to some percentage of the `CppRef::ptr`s.

The "pointer" inside here really should be thought of as an opaque token. It might not even be a pointer. [Here's an example where it would be some index into a lookup table](https://github.com/rust-lang/rust/issues/103516#issuecomment-1790856267).

I think what you're hinting at here is a deeper concern: unless we _can_ document reasonable invariants for `CppRef::ptr`, then we should not allow that pointer to exist in Rust, because neither the compiler (in safe code) nor the human (in unsafe code) can guarantee those invariants are met. I can see that point of view. But this is to say that we should not have interop with other languages in Rust, because they intrinsically have weaker guarantees than Rust. So long as we accept that interop with other languages is a necessity, then we should allow the Rust user to build safer abstractions to reduce the practical risk of the unsafety. That's what this is all about.

Reply:
I don't think actually documenting them would distract from the RFC at all. An important part of an RFC is to show how the RFC would be used on real code, and for using this RFC in real code, having such a safety invariant is absolutely crucial! The entire RFC rests in having raw pointers in private fields that we know extra things about due to an invariant, and relying on that invariant to provide a safe API.

Reply:
Note to other readers: in another comment thread we discovered that this RFC wasn't sufficiently clear that we're mostly aiming at _auto generated bindings_. Over the coming days I'll update the RFC to make that clearer, which should also end up resolving this comment thread by more accurately describing the invariants which are, and are not, realistically possible for such auto-generated bindings. I'll resolve this comment thread too once I've made that update.

Reply:
I hope this is addressed by 69b1edfc07a8926e7a60638a7a9cb336f2cb86ef. I'm going to mark this conversation thread as resolved, but we can continue to discuss on [the related comment thread](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1380344160) if it's not sorted out.